### DISCUSSION TITLE
writing custom .net runtime host

### DISCUSSION THREAD

**Discussion by shivamdabas** [dotnet/runtime#70101] [Q&A]
Hi,
I implemented a custom .net runtime host using the github code provided by Microsoft  but I'm facing a peculiar issue.
I have implemented ODBC driver dll which is in C++ and and a custom .net runtime host which is calling .net core 5.0 dll to make use of Microsoft Authentication Libarary in .net but the problem that I'm facing is when this ODBC driver dll is getting called from another exe developed using .net 5.0 but with System.Data.Odbc nuget package  , my program is crashing in custom .net runtime host trying to load the .net core and initialize the function pointer. 

My code for loading .net core

```
 load_assembly_and_get_function_pointer_fn get_dotnet_load_assembly(const char_t* config_path)
    {
        // Load .NET Core
        void* load_assembly_and_get_function_pointer = nullptr;
        hostfxr_handle cxt = nullptr;
        int rc = init_fptr(config_path, nullptr, &cxt);   // failing here 
        if (rc != 0 || cxt == nullptr)
        {
            TRACE1("Init failed: %d ", rc);
            close_fptr(cxt);
            return nullptr;
        }

        // Get the load assembly function pointer
        rc = get_delegate_fptr(
            cxt,
            hdt_load_assembly_and_get_function_pointer,
			&load_assembly_and_get_function_pointer);
		if (rc != 0 || load_assembly_and_get_function_pointer == nullptr)
		{
			TRACE1("Get delegate failed: %d ", rc);
		}

        close_fptr(cxt);
        return (load_assembly_and_get_function_pointer_fn)load_assembly_and_get_function_pointer;
    }
```
The thing is if I call my ODBC driver dll from exact .net framework code it works fine.

Any help would be appreciated as I'm new to .net

**Comment by vitek-karas** [comment]
How does it fail (at least error code please).
What is the app you're loading it into - is it framework dependent or self-contained? Anything else potentially special about it?

**Reply by shivamdabas** [reply]
> How does it fail (at least error code please). What is the app you're loading it into - is it framework dependent or self-contained? Anything else potentially special about it?

Thanks for your reply, I'm calling "init_fptr" function which is an alias for "hostfxr_initialize_for_runtime_config_fn" from "hostfxr.h" file where it is failing which is returning 1 as error code.

This is my .net 5.0 framework dependent  code to call my ODBC driver dll developed in C++ with custom .net runtime host to load .net 5.0 framework dependent dll

Interestingly if I use .net framework instead of .net core with the below code, everything works fine

```
using System;
using System.Data.Odbc;

namespace mssqlTestApp
{
    class Program
    {
        static void Main(string[] args)
        {
            if (args.Length < 1)
                throw new Exception("Connection string is not specified.");
            var connectionString = args[0];

            OdbcConnection connection = null;
            try
            {
                connection = new OdbcConnection(connectionString);  // it is calling my ODBC driver with custom .net runtimehost and failing 
                                                                                                          // here 

                try
                {
                    connection.Open();
                    Console.WriteLine("Connection established successfully.");
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"Exception occured while opening connection: {ex.Message}");
                }
            }
            catch (Exception e)
            {
                Console.WriteLine(e);
            }
            finally
            {
                connection?.Close();
            }
        }
    }
}
```

**Reply by vitek-karas** [reply]
1 is not an error - it's a different success. In this case it means that there was already a runtime in the process and the initialize attached to it. https://github.com/dotnet/runtime/blob/main/docs/design/features/native-hosting.md#initialize-host-context. You can check that it did return a context handle.

https://github.com/dotnet/runtime/blob/main/docs/design/features/host-error-codes.md see `Success_HostAlreadyInitialized`.

**Reply by shivamdabas** [reply]
> 1 is not an error - it's a different success. In this case it means that there was already a runtime in the process and the initialize attached to it. https://github.com/dotnet/runtime/blob/main/docs/design/features/native-hosting.md#initialize-host-context. You can check that it did return a context handle.
> 
> https://github.com/dotnet/runtime/blob/main/docs/design/features/host-error-codes.md see `Success_HostAlreadyInitialized`.

Thanks for these references, As I can see it did return a context handle in this case. I will check with that handle itself and update here

**Reply by shivamdabas** [reply]
Thanks it worked on windows but on linux I'm getting error code as 0x80008092 which is LibHostInvalidArgs, I'm not able able to point where it might be wrong as code is same on both linux and windows. Any insight on that?

**Reply by vitek-karas** [reply]
Set `COREHOST_TRACE=1` and rerun - it should print out some detailed info into the stderr. It's not 100% though, some failure will not print out much - but it should at least help locale where it's failing.

**Reply by shivamdabas** [reply]
I'm able to solve the issue, was a minor one needed to run through the executable instead of dll
Thanks for your help
